### PR TITLE
Supporting `order` arg to `asnumpy`

### LINKED ISSUE #2568: `dpnp.asnumpy` does not honor `order` arg
Cross-linking with the [issue](https://github.com/IntelPython/dpnp/issues/2884#issuecomment-3222990884) from dpctl for supporting `order` arg in `dpt.asnumpy`

Reproducer:
```python
import numpy as np
import dpnp as dp

print("dpnp version:", getattr(dp, "__version__", "unknown"))

# Build the same data as in the larger code
ctr_offsets_slice = [
    dp.array([0, 0, 0], dtype=np.int32),
    dp.array([3, 3, 3], dtype=np.int32),
    dp.array([5, 5, 5], dtype=np.int32),
    dp.array([7, 7, 7], dtype=np.int32),
    dp.array([10, 10, 10], dtype=np.int32),
    dp.array([11, 11, 11], dtype=np.int32),
    dp.array([12, 12, 12], dtype=np.int32),
]

# 3a: stack -> dpnp_array (should be C-contiguous)
temp1 = dp.stack(ctr_offsets_slice)

# 3b: transpose -> dpnp_array view (not C-contiguous)
temp2 = dp.stack(ctr_offsets_slice).T

# 3c: device->host with "order='C'"
# EXPECTED (per docstring "works exactly like numpy.asarray"): C-contiguous NumPy array
# ACTUAL: order is ignored for dpnp_array, result keeps non-C layout
temp3 = dp.asnumpy(temp2, order='C')

print("3a. Testing : temp1: ", bool(temp1.flags['C_CONTIGUOUS']), type(temp1), len(temp1), temp1)
print("3b. Testing : temp2: ", bool(temp2.flags['C_CONTIGUOUS']), type(temp2), len(temp2), temp2)
print("3c. Testing : temp3: ", bool(temp3.flags['C_CONTIGUOUS']), type(temp3), len(temp3), temp3)

# Programmatic check to make the failure obvious:
if not temp3.flags['C_CONTIGUOUS']:
    print("\nBUG: dpnp.asnumpy(dpnp_array, order='C') returned a non-C-contiguous NumPy array.")
    print("     strides:", temp3.strides, "| shape:", temp3.shape)
```

output from DPNP+DPCTL
```bash
dpnp version: 0.19.0dev3+8.gac795ab9.dirty
3a. Testing : temp1:  True <class 'dpnp.dpnp_array.dpnp_array'> 7 [[ 0  0  0]
 [ 3  3  3]
 [ 5  5  5]
 [ 7  7  7]
 [10 10 10]
 [11 11 11]
 [12 12 12]]
3b. Testing : temp2:  False <class 'dpnp.dpnp_array.dpnp_array'> 3 [[ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]]
3c. Testing : temp3:  False <class 'numpy.ndarray'> 3 [[ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]]

BUG: dpnp.asnumpy(dpnp_array, order='C') returned a non-C-contiguous NumPy array.
     strides: (4, 12) | shape: (3, 7)
```

output from cupy
```bash
cupy version: 13.4.1
3a. Testing : temp1:  True <class 'cupy.ndarray'> 7 [[ 0  0  0]
 [ 3  3  3]
 [ 5  5  5]
 [ 7  7  7]
 [10 10 10]
 [11 11 11]
 [12 12 12]]
3b. Testing : temp2:  False <class 'cupy.ndarray'> 3 [[ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]]
3c. Testing : temp3:  True <class 'numpy.ndarray'> 3 [[ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]
 [ 0  3  5  7 10 11 12]]
```

### PR DESCRIPTION
**Pull request by abagusetty**
Fixed `dpnp.asnumpy` and `dpnp.ndarray.asnumpy` ignoring the `order` arg
Fixes: https://github.com/IntelPython/dpnp/issues/2568

- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, reproducer or referred to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [x] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [x] Have you added documentation for your changes, if necessary?
- [x] Have you added your changes to the changelog?

### CODE REVIEW

File: dpnp/tensor/_copy_utils.py [hunk 1]
```diff
@@ -605,12 +608,16 @@ def asnumpy(usm_ary):
     Args:
         usm_ary (usm_ndarray):
             Input array
+        order (``"C"``, ``"F"``, ``"A"``, ``"K"``):
```
Comment (antonwolfy) [inline dpnp/tensor/_copy_utils.py]:
```suggestion
        order (``"C"``, ``"F"``, ``"A"``, ``"K"``, optional):
```

File: dpnp/tensor/_copy_utils.py [hunk 2]
```diff
@@ -51,12 +51,12 @@
...
 
 
-def _copy_to_numpy(ary):
+def _copy_to_numpy(ary, order="K"):
     if not isinstance(ary, dpt.usm_ndarray):
         raise TypeError(f"Expected dpnp.tensor.usm_ndarray, got {type(ary)}")
     if ary.size == 0:
         # no data needs to be copied for zero sized array
-        return np.ndarray(ary.shape, dtype=ary.dtype)
+        return np.ndarray(ary.shape, dtype=ary.dtype, order=order)
```
Comment (antonwolfy) [inline dpnp/tensor/_copy_utils.py]: I wonder if it is allowed to pass `K` and `A` values here, because based on the [documentation](https://numpy.org/doc/stable/reference/generated/numpy.ndarray.html) it's limited to `C` and `F` only:
> {‘C’, ‘F’}, optional

Comment (antonwolfy) [reply dpnp/tensor/_copy_utils.py]: Any `size == 0` array reports `C_CONTIGUOUS = True` and `F_CONTIGUOUS = True` regardless of `order`.
It'd make sense to leave as original:
```suggestion
        return np.ndarray(ary.shape, dtype=ary.dtype)
```
but place a comment

File: CHANGELOG.md [hunk 3]
```diff
@@ -54,6 +54,7 @@ This release is compatible with NumPy 2.5.
 * Fixed `dpnp.linalg.svd(..., hermitian=True)` returning a non-unitary `vh` for singular input arrays due to a zero sign appearing [#2954](https://github.com/IntelPython/dpnp/pull/2954)
 * Fixed scalar conversion of size-one `dpnp.tensor.usm_ndarray` (e.g. `int()`, `float()`, indexing) which failed with NumPy 2.5 after the in-place `ndarray.shape` assignment was deprecated [#2958](https://github.com/IntelPython/dpnp/pull/2958)
 * Fixed `dpnp.mgrid` and `dpnp.ogrid` to return consistent results between single-slice and tuple-of-slices syntax when the step is a complex number with a non-integer magnitude (e.g. `2.5j`) [#2971](https://github.com/IntelPython/dpnp/pull/2971)
+* Fixed `dpnp.asnumpy` and `dpnp.ndarray.asnumpy` ignoring the `order` keyword, which caused a non-contiguous source array to be returned with a non-contiguous layout even when `order="C"` was requested [#2884](https://github.com/IntelPython/dpnp/issues/2884)
```
Comment (antonwolfy) [inline CHANGELOG.md]: Seems wrong PR number:
```suggestion
* Fixed `dpnp.asnumpy` and `dpnp.ndarray.asnumpy` ignoring the `order` keyword, which caused a non-contiguous source array to be returned with a non-contiguous layout even when `order="C"` was requested [#2980](https://github.com/IntelPython/dpnp/issues/2980)
```

File: dpnp/tensor/_copy_utils.py (hunk 1 above)
Comment (antonwolfy) [reply dpnp/tensor/_copy_utils.py]: I believe `None` value is also support for `tensor.asnumpy`

File: dpnp/tests/test_ndarray.py [hunk 4]
```diff
@@ -143,6 +143,59 @@ def test_roundtrip_binary_str(self, order):
...
+    def test_method_default_order_is_c(self):
+        # the array method matches ``cupy.ndarray.get`` and defaults to "C"
+        a = self._non_c_contiguous_array()
+        result = a.asnumpy()
+        assert result.flags["C_CONTIGUOUS"]
+
+    def test_method_order_k_keeps_strides(self):
+        # explicit "K" keeps the strides of the source as closely as possible
+        a = self._non_c_contiguous_array()
+        result = a.asnumpy(order="K")
```
Comment (antonwolfy) [inline dpnp/tests/test_ndarray.py:203]: missing test for `order="A"` and `order=None`

File: dpnp/tests/test_ndarray.py [hunk 5]
```diff
@@ -143,6 +143,59 @@ def test_roundtrip_binary_str(self, order):
...
+        assert not result.flags["C_CONTIGUOUS"]
+
```
Comment (antonwolfy) [inline dpnp/tests/test_ndarray.py:205]: Messing test with zero-sized ndarray

**Comment by abagusetty** [comment]
@antonwolfy Thanks for taking time. Addressed the above suggestions

**Comment by ndgrigorian** [comment]
@abagusetty 
while this works in some cases, in other edge cases, this will not work, especially in terms of keeping `"K"` order:
```
In [1]: import numpy as np

In [2]: import dpnp as dnp

In [3]: x = np.arange(10)

In [4]: x_np = np.arange(10)

In [5]: x = dnp.arange(10)

In [6]: x
Out[6]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [7]: x_np
Out[7]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [8]: x = x[::-1]

In [9]: x_np = x_np[::-1]

In [10]: dnp.asnumpy(x)
Out[10]: array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

In [11]: _.strides
Out[11]: (8,)

In [12]: np.asarray(x_np)
Out[12]: array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

In [13]: _.strides
Out[13]: (-8,)
```

the issue is that there are no kernels for copying to strided host memory, only contiguous. So in such a case as this, we get a C-contiguous copy back. With cases like F-contiguous memory, `order="K"` will preserve it, because we can copy to F-contiguous host memory, as it is contiguous.

A proper solution requires new kernels. At minimum, the limitation should be documented.

File: dpnp/tensor/_copy_utils.py [hunk 6]
```diff
@@ -67,13 +70,16 @@ def _copy_to_numpy(ary):
...
+    result = np.ndarray(
         ary.shape,
         dtype=ary.dtype,
         buffer=h,
         strides=strides_bytes,
         offset=offset,
     )
+    # apply the requested memory layout; ``"K"`` preserves the strides of the
+    # source array as closely as possible and is the default
+    return np.asarray(result, order=order)
```
Comment (ndgrigorian) [inline dpnp/tensor/_copy_utils.py:82]: part of the reason is that this will preserve the strides of the source result array, which are not necessarily correct per NumPy's `order="K"` semantics themselves

**Comment by ndgrigorian** [comment]
@abagusetty 
it seems `order="K"` by default wasn't changed for `dnp.asnumpy`, so that observation was incorrect!

```
In [10]: dpt.asnumpy(x_dpt)
Out[10]: array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

In [11]: _.strides
Out[11]: (-8,)

In [12]: dnp.asnumpy(x_dnp, order="K")
Out[12]: array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])

In [13]: _.strides
Out[13]: (-8,)
```
I will look for other edge cases, but if this works in this case, I don't see a likely issue otherwise

**Comment by abagusetty** [comment]
> @abagusetty it seems `order="K"` by default wasn't changed for `dnp.asnumpy`, so that observation was incorrect!
> 
> ```
> In [10]: dpt.asnumpy(x_dpt)
> Out[10]: array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])
> 
> In [11]: _.strides
> Out[11]: (-8,)
> 
> In [12]: dnp.asnumpy(x_dnp, order="K")
> Out[12]: array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])
> 
> In [13]: _.strides
> Out[13]: (-8,)
> ```
> 
> I will look for other edge cases, but if this works in this case, I don't see a likely issue otherwise

@ndgrigorian Thanks for the feedback. I've added a negative-stride case too into the tests

```bash
>>> import numpy as np
>>> import dpnp as dnp
>>> import dpnp.tensor as dpt
>>> x_dnp = dnp.arange(10)[::-1]
>>> x_dpt = dnp.get_usm_ndarray(x_dnp)
>>> print(dpt.asnumpy(x_dpt))
[9 8 7 6 5 4 3 2 1 0]
>>> print(dpt.asnumpy(x_dpt).strides)
(-8,)
>>> print(dnp.asnumpy(x_dnp, order="K"))
[9 8 7 6 5 4 3 2 1 0]
>>> print(dnp.asnumpy(x_dnp, order="K").strides)
(-8,)
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -57,6 +57,7 @@ This release is compatible with NumPy 2.5.
 * Fixed `dpnp.linalg.svd(..., hermitian=True)` returning a non-unitary `vh` for singular input arrays due to a zero sign appearing [#2954](https://github.com/IntelPython/dpnp/pull/2954)
 * Fixed scalar conversion of size-one `dpnp.tensor.usm_ndarray` (e.g. `int()`, `float()`, indexing) which failed with NumPy 2.5 after the in-place `ndarray.shape` assignment was deprecated [#2958](https://github.com/IntelPython/dpnp/pull/2958)
 * Fixed `dpnp.mgrid` and `dpnp.ogrid` to return consistent results between single-slice and tuple-of-slices syntax when the step is a complex number with a non-integer magnitude (e.g. `2.5j`) [#2971](https://github.com/IntelPython/dpnp/pull/2971)
+* Fixed `dpnp.asnumpy` and `dpnp.ndarray.asnumpy` ignoring the `order` keyword, which caused a non-contiguous source array to be returned with a non-contiguous layout even when `order="C"` was requested [#2980](https://github.com/IntelPython/dpnp/issues/2980)
```
Comment (antonwolfy) [inline CHANGELOG.md]:
```suggestion
* Fixed `dpnp.asnumpy` and `dpnp.ndarray.asnumpy` ignoring the `order` keyword, which caused a non-contiguous source array to be returned with a non-contiguous layout even when `order="C"` was requested [#2980](https://github.com/IntelPython/dpnp/pull/2980)
```

**Review by ndgrigorian — APPROVED** [review]
with negative strides working, this LGTM

**Review by antonwolfy — APPROVED** [review]
Thank you @abagusetty!
